### PR TITLE
Expand pushdown support with more comparison and logical operations

### PR DESCRIPTION
**Pull request by AdamGS**
Add support for more comparisons and add recursive validation logic for expressions.

### CODE REVIEW

File: vortex-datafusion/src/eval.rs [hunk 1]
```diff
@@ -14,34 +14,35 @@ impl ExpressionEvaluator {
...
+                    DFOperator::Lt => compare(&lhs, &rhs, Operator::Lt),
+                    DFOperator::LtEq => compare(&lhs, &rhs, Operator::Lte),
+                    DFOperator::NotEq => compare(&lhs, &rhs, Operator::NotEq),
                     _ => vortex_bail!("{} is an unsupported operator", expr.op),
                 }
             }
             Expr::Column(col) => {
                 // TODO(adamg): Use variant trait once its merged
-                let array = array.clone().into_struct()?;
+                let array = array.into_struct()?;
```
Comment (gatesn) [inline vortex-datafusion/src/eval.rs]: `array.with_dyn(|a| a.as_struct_array()).ok_or(|| "can't grab a column from a non-struct array blah blah")`

File: vortex-datafusion/src/lib.rs [hunk 2]
```diff
@@ -288,33 +297,76 @@ fn make_filter_then_take_plan(
...
+fn can_be_pushed_down(expr: &Expr) -> DFResult<bool> {
+    fn is_supported(expr: &Expr) -> bool {
+        match expr {
+            Expr::BinaryExpr(expr)
+                if expr.op.is_logic_operator() || SUPPORTED_BINARY_OPS.contains(&expr.op) =>
+            {
+                let lhs = expr.left.as_ref();
+                let rhs = expr.right.as_ref();
+
+                matches!(
```
Comment (gatesn) [inline vortex-datafusion/src/lib.rs]: Can we just recurse here?

File: vortex-datafusion/src/eval.rs [hunk 3]
```diff
@@ -14,34 +14,35 @@ impl ExpressionEvaluator {
...
+                        let rhs = rhs.into_bool()?;
                         let buffer = &lhs.boolean_buffer() | &rhs.boolean_buffer();
                         Ok(BoolArray::from(buffer).into_array())
                     }
-                    DFOperator::Eq => {
-                        let lhs = ExpressionEvaluator::eval(array.clone(), lhs)?;
-                        let rhs = ExpressionEvaluator::eval(array.clone(), rhs)?;
-                        compare(&lhs, &rhs, Operator::Eq)
-                    }
+                    DFOperator::Eq => compare(&lhs, &rhs, Operator::Eq),
```
Comment (gatesn) [inline vortex-datafusion/src/eval.rs:33]: Can we define a `TryFrom<DFOperator> for Operator`? And then use that to check for support?

**Review by gatesn — CHANGES_REQUESTED** [review]

File: vortex-datafusion/src/eval.rs (hunk 1 above)
Comment (AdamGS) [reply vortex-datafusion/src/eval.rs]: Resolved with some help from @robert3005

File: vortex-datafusion/src/eval.rs (hunk 3 above)
Comment (AdamGS) [reply vortex-datafusion/src/eval.rs:33]: `And` and `Or` don't really fit into how `Operator` is designed right now with functions like `to_predicate` or even `inverse`

File: vortex-datafusion/src/eval.rs [hunk 4]
```diff
@@ -14,39 +13,38 @@ impl ExpressionEvaluator {
...
                     _ => vortex_bail!("{} is an unsupported operator", expr.op),
                 }
             }
-            Expr::Column(col) => {
-                // TODO(adamg): Use variant trait once its merged
-                let array = array.clone().into_struct()?;
+            Expr::Column(col) => array.with_dyn(|a| {
                 let name = col.name();
-                array
+                a.as_struct_array_unchecked()
```
Comment (gatesn) [inline vortex-datafusion/src/eval.rs]: Why unchecked?

**Review by gatesn — APPROVED** [review]

**Pull request merged**